src/main/java/uk/gov/ons/census/casesvc/messaging/ManagedMessageRecoverer.java

```diff
@@ -0,0 +1,224 @@
+
+      peekMessage(reportResult, messageHash, rawMessageBody);
+      logMessage(
+          reportResult, listenerExecutionFailedException.getCause(), messageHash, rawMessageBody);
+
+      // Send the bad message to an exchange where it'll be retried at some future point in time
+      rabbitTemplate.send(delayExchangeName, queueName, message);
+    } else {
+      // Very unlikely that this'd happen but let's log it anyway
+      log.error("Unexpected exception has occurred", throwable);
+    }
+  }
+
+  private ExceptionReportResponse getExceptionReportResponse(
+      ListenerExecutionFailedException listenerExecutionFailedException, String messageHash) {
+    ExceptionReportResponse reportResult = null;
+    try {
+      reportResult =
+          exceptionManagerClient.reportException(
+              messageHash,
+              serviceName,
+              queueName,
+              listenerExecutionFailedException.getCause().getCause());
+    } catch (Exception exceptionManagerClientException) {
+      log.with("reason", exceptionManagerClientException.getMessage())
+          .warn(
+              "Could not report to Exception Manager. There will be excessive logging until resolved");
+    }
+    return reportResult;
+  }
+
+  private boolean skipMessage(
+      ExceptionReportResponse reportResult,
+      String messageHash,
+      byte[] rawMessageBody,
+      ListenerExecutionFailedException listenerExecutionFailedException,
+      Message message) {
+    if (reportResult == null || !reportResult.isSkipIt()) {
+      return false;
+    }
+
+    boolean result = false;
+
+    // Make damn certain that we have a copy of the message before skipping it
+    try {
+      SkippedMessage skippedMessage = new SkippedMessage();
+      skippedMessage.setMessageHash(messageHash);
+      skippedMessage.setMessagePayload(rawMessageBody);
+      skippedMessage.setService(serviceName);
+      skippedMessage.setQueue(queueName);
+      skippedMessage.setContentType(
+          listenerExecutionFailedException
+              .getFailedMessage()
+              .getMessageProperties()
+              .getContentType());
+      skippedMessage.setHeaders(
+          listenerExecutionFailedException.getFailedMessage().getMessageProperties().getHeaders());
+      skippedMessage.setRoutingKey(
+          listenerExecutionFailedException
+              .getFailedMessage()
+              .getMessageProperties()
+              .getReceivedRoutingKey());
+      exceptionManagerClient.storeMessageBeforeSkipping(skippedMessage);
+      result = true;
+    } catch (Exception exceptionManagerClientException) {
+      log.with("message_hash", messageHash)
+          .warn(
+              "Unable to store a copy of the message. Will NOT be skipping",
+              exceptionManagerClientException);
+    }
+
+    // Check if OK and the message is stored... then we can go ahead and quarantine
+    if (result) {
+      result = false; // The next bit might go wrong
```

Comment:
not sure about setting this result value here? I presume if there's an Exception send the message the 'result' won't get returned anyway?

Reply:
yeah I'm just lazily re-using the boolean variable so I don't have too many return statements and wotnot